### PR TITLE
[Feature]: SimMIM supported

### PR DESCRIPTION
**Pull request by YuanLiuuuuuu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: configs/benchmarks/classification/_base_/models/swin-base.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# model settings
+
+custom_imports = dict(imports='mmcls.models', allow_failed_imports=False)
+
+model = dict(
+    type='ClassificationForCls',
```
Comment (fangyixiao18) [inline configs/benchmarks/classification/_base_/models/swin-base.py]: Maybe change name to `ClassifierWrapper`?

File: mmselfsup/models/necks/simmim_neck.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch.nn as nn
+from mmcv.runner import BaseModule
+
+from ..builder import NECKS
+
+
+@NECKS.register_module()
+class SimMIMNeck(BaseModule):
+
```
Comment (fangyixiao18) [inline mmselfsup/models/necks/simmim_neck.py:12]: missing docstring

File: mmselfsup/models/backbones/__init__.py [hunk 3]
```diff
@@ -3,9 +3,10 @@
 from .mim_cls_vit import MIMVisionTransformer
 from .resnet import ResNet, ResNetV1d
 from .resnext import ResNeXt
+from .simmim_swin import SwinForSimMIM
 from .vision_transformer import VisionTransformer
 
 __all__ = [
     'ResNet', 'ResNetV1d', 'ResNeXt', 'MAEViT', 'MIMVisionTransformer',
-    'VisionTransformer'
+    'VisionTransformer', 'SwinForSimMIM'
```
Comment (fangyixiao18) [inline mmselfsup/models/backbones/__init__.py]: The name format is not consistent with `MAEViT`

File: mmselfsup/models/backbones/simmim_swin.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    def _init_weights(self, m):
+        if isinstance(m, nn.Linear):
+            trunc_normal_(m.weight, std=.02)
+            if isinstance(m, nn.Linear) and m.bias is not None:
+                nn.init.constant_(m.bias, 0)
+        elif isinstance(m, nn.LayerNorm):
+            nn.init.constant_(m.bias, 0)
+            nn.init.constant_(m.weight, 1.0)
+
+    def forward(self, x: torch.Tensor,
```
Comment (fangyixiao18) [inline mmselfsup/models/backbones/simmim_swin.py:107]: add some docstring and comment for forward function

File: mmselfsup/core/optimizer/constructor.py [hunk 5]
```diff
@@ -10,11 +10,10 @@
 class DefaultOptimizerConstructor:
     """Rewrote default constructor for optimizers. By default each parameter
     share the same optimizer settings, and we provide an argument
-    ``paramwise_cfg`` to specify parameter-wise settings. It is a dict and may
-    contain the following fields:
+    ``paramwise_cfg`` to specify parameter-wise settings and set layer-wise
+    learning rate. It is a dict and may contain the following fields:
```
Comment (fangyixiao18) [inline mmselfsup/core/optimizer/constructor.py]: this part is not completed?

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

File: mmselfsup/models/algorithms/classification_cls.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+from mmcls.models import ImageClassifier
+from mmcv.runner import auto_fp16
+
+from ..builder import ALGORITHMS
+
+
+@ALGORITHMS.register_module()
+class ClassificationForCls(ImageClassifier):
```
Comment (ZwwWayne) [inline mmselfsup/models/algorithms/classification_cls.py]: ClassificationForCls -> MMClsImageClassifierWrapper

File: mmselfsup/models/backbones/__init__.py (hunk 3 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/models/backbones/__init__.py]: Keep consistency with the official code

**Comment by fangyixiao18** [comment]
I found that the files in folder `docs/en/algorithms` are outdated, thus I think it is better to create a new PR to update them. And your latest two commits are part of that PR.

**Comment by fangyixiao18** [comment]
Metafile and SimMIM algorithm readme are missing.

**Comment by YuanLiuuuuuu** [comment]
> I found that the files in folder `docs/en/algorithms` are outdated, thus I think it is better to create a new PR to update them. And your latest two commits are part of that PR.

I think there is no need to update them separately, and the outdated problem will be fixed synchronouly with this PR, since this PR will be merged soon.

**Comment by fangyixiao18** [comment]
> I think there is no need to update them separately, and the outdated problem will be fixed synchronouly with this PR, since this PR will be merged soon.

OK

File: configs/selfsup/simmim/README.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
...image modeling task learn good representations, we systematically study the major components in our framework, and find that simple designs of each component have revealed very strong representation learning performance: 1) random masking of the input image with a moderately large masked patch size (e.g., 32) makes a strong pre-text task; 2) predicting raw pixels of RGB values by direct regression performs no worse than the patch classification approaches with complex designs; 3) the prediction head can be as light as a linear layer, with no worse performance than heavier ones. Using ViT-B, our approach achieves 83.8% top-1 fine-tuning accuracy on ImageNet-1K by pre-training also on this dataset, surpassing previous best approach by +0.6%. When applied on a larger model of about 650 million parameters, SwinV2H, it achieves 87.1% top-1 accuracy on ImageNet-1K using only ImageNet-1K data. We also leverage this approach to facilitate the training of a 3B model (SwinV2-G), that by 40× less data than that in previous practice, we achieve the state-of-the-art on four representative vision benchmarks. The code and models will be publicly available at https: //github.com/microsoft/SimMIM .
```
Comment (ZwwWayne) [inline configs/selfsup/simmim/README.md]: typo

File: docs/en/algorithms/mae.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
+# MAE
```
Comment (ZwwWayne) [inline docs/en/algorithms/mae.md]: We should split these irrelevant contents into another PR.

File: mmselfsup/core/optimizer/constructor.py [hunk 9]
```diff
@@ -41,13 +45,29 @@ def __init__(self, optimizer_cfg, paramwise_cfg=None):
                             f'but got {type(optimizer_cfg)}')
         self.optimizer_cfg = optimizer_cfg
         self.paramwise_cfg = {} if paramwise_cfg is None else paramwise_cfg
+        self.layer_decay = self.optimizer_cfg.pop('layer_decay', 0.0)
```
Comment (ZwwWayne) [inline mmselfsup/core/optimizer/constructor.py]: 1. I suggest inheriting another TransformerFinetuneConstructor rather than update the original defaultconstructor.
2. TransformerFinetuneConstructor could support change layerwise decay of both swin and vit.

File: mmselfsup/datasets/pipelines/transforms.py [hunk 10]
```diff
@@ -17,6 +18,49 @@
         PIPELINES.register_module(m[1])
 
 
+@PIPELINES.register_module()
+class BlockMaskGen(object):
```
Comment (ZwwWayne) [inline mmselfsup/datasets/pipelines/transforms.py]: BlockwiseMaskGenerator might be better

File: mmselfsup/models/backbones/simmim_swin.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+import torch
+import torch.nn as nn
+from mmcls.models import SwinTransformer
+from mmcv.cnn.utils.weight_init import trunc_normal_
+
+from ..builder import BACKBONES
+
+
+@BACKBONES.register_module()
+class SwinForSimMIM(SwinTransformer):
```
Comment (ZwwWayne) [inline mmselfsup/models/backbones/simmim_swin.py]: SimMIMSwinTransformer

File: configs/selfsup/simmim/README.md [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+| :------: | :-------------: | :---------------: | :-------------------------------------------------: | :---------------------------------------------------------------------------------------: | :-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
+| Swin-Base |       100       |       82.9        | [config](./simmim_swin-base_16xb128-coslr-100e_in1k-192) | [config](../../benchmarks/classification/imagenet/swin-base_ft-8xb256-coslr-100e_in1k) | [model](https://download.openmmlab.com/mmselfsup/simmim/simmim_swin-base_16xb128-coslr-100e_in1k-192_20220316-1d090125.pth) &#124; [log](https://download.openmmlab.com/mmselfsup/simmim/simmim_swin-base_16xb128-coslr-100e_in1k-192_20220316-1d090125.log.json) |
+
+
+## Citation
+
+```bibtex
+@inproceedings{xie2021simmim,
```
Comment (Jiahao000) [inline configs/selfsup/simmim/README.md:30]: Should be `xie2022simmim`?

File: docs/en/algorithms/simmim.md [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
...image modeling task learn good representations, we systematically study the major components in our framework, and find that simple designs of each component have revealed very strong representation learning performance: 1) random masking of the input image with a moderately large masked patch size (e.g., 32) makes a strong pre-text task; 2) predicting raw pixels of RGB values by direct regression performs no worse than the patch classification approaches with complex designs; 3) the prediction head can be as light as a linear layer, with no worse performance than heavier ones. Using ViT-B, our approach achieves 83.8% top-1 fine-tuning accuracy on ImageNet-1K by pre-training also on this dataset, surpassing previous best approach by +0.6%. When applied on a larger model of about 650 million parameters, SwinV2H, it achieves 87.1% top-1 accuracy on ImageNet-1K using only ImageNet-1K data. We also leverage this approach to facilitate the training of a 3B model (SwinV2-G), that by 40× less data than that in previous practice, we achieve the state-of-the-art on four representative vision benchmarks. The code and models will be publicly available at https: //github.com/microsoft/SimMIM .
```
Comment (Jiahao000) [inline docs/en/algorithms/simmim.md]: Please fix typos in the abstract.

File: docs/zh_cn/algorithms/simmim.md [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
...image modeling task learn good representations, we systematically study the major components in our framework, and find that simple designs of each component have revealed very strong representation learning performance: 1) random masking of the input image with a moderately large masked patch size (e.g., 32) makes a strong pre-text task; 2) predicting raw pixels of RGB values by direct regression performs no worse than the patch classification approaches with complex designs; 3) the prediction head can be as light as a linear layer, with no worse performance than heavier ones. Using ViT-B, our approach achieves 83.8% top-1 fine-tuning accuracy on ImageNet-1K by pre-training also on this dataset, surpassing previous best approach by +0.6%. When applied on a larger model of about 650 million parameters, SwinV2H, it achieves 87.1% top-1 accuracy on ImageNet-1K using only ImageNet-1K data. We also leverage this approach to facilitate the training of a 3B model (SwinV2-G), that by 40× less data than that in previous practice, we achieve the state-of-the-art on four representative vision benchmarks. The code and models will be publicly available at https: //github.com/microsoft/SimMIM .
```
Comment (Jiahao000) [inline docs/zh_cn/algorithms/simmim.md]: Same typos here

**Review by Jiahao000 — CHANGES_REQUESTED** [review]

File: configs/selfsup/simmim/README.md (hunk 12 above)
Comment (YuanLiuuuuuu) [reply configs/selfsup/simmim/README.md:30]: This bibtex is token from the [official repo](https://github.com/microsoft/SimMIM)

**Review by fangyixiao18 — APPROVED** [review]

File: mmselfsup/models/heads/simmim_head.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+from mmcv.runner import BaseModule
+from torch.nn import functional as F
+
+from ..builder import HEADS
+
+
+@HEADS.register_module()
+class SimMIMPretrainHead(BaseModule):
```
Comment (Jiahao000) [inline mmselfsup/models/heads/simmim_head.py]: Better to change to `SimMIMHead` as there is no finetune head (like `MoCoV3Head`)?

**Review by Jiahao000 — CHANGES_REQUESTED** [review]

File: configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py [hunk 16]
```diff
@@ -45,7 +45,8 @@
         'pos_embed': dict(weight_decay=0.),
         'cls_token': dict(weight_decay=0.)
     },
-    constructor='MAEFtOptimizerConstructor',
+    constructor='TransformerFinetuneConstructor',
+    model_type='vit',
     layer_decay=0.65)
```
Comment (fangyixiao18) [inline configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py:51]: rename the file, `vit-b-p16` -> `vit-base-p16`?

**Review by Jiahao000 — APPROVED** [review]

File: docs/en/index.rst [hunk 17]
```diff
@@ -40,8 +40,7 @@ Welcome to MMSelfSup's documentation!
    algorithms/simclr.md
    algorithms/ss.md
    algorithms/swav.md
-   algorithms/mocov3.md
-   algorithms/mae.md
```
Comment (fangyixiao18) [inline docs/en/index.rst:44]: recover the deletion?

File: docs/zh_cn/index.rst [hunk 18]
```diff
@@ -42,8 +42,7 @@ Welcome to MMSelfSup's documentation!
    algorithms/simclr.md
    algorithms/ss.md
    algorithms/swav.md
-   algorithms/mocov3.md
-   algorithms/mae.md
```
Comment (fangyixiao18) [inline docs/zh_cn/index.rst:46]: the same
